src/app/qgsbookmarks.cpp

```diff
@@ -375,3 +380,314 @@ void QgsBookmarks::exportToXML()
+    }
+    else
+    {
+      return qgisTableModel.data( this->index( 0, index.column() ), role );
+    }
+  }
+}
+
+bool QgsMergedBookmarksTableModel::setData( const QModelIndex& index, const QVariant& value, int role )
+{
+  // is project or QGIS
+  if ( index.column() == qgisTableModel.columnCount() )
+  {
+    if ( index.row() < qgisTableModel.rowCount() )
+    {
+      moveBookmark( qgisTableModel, projectTableModel, index.row() );
+      treeView->scrollTo( this->index( rowCount() - 1, 1 ) );
+      treeView->setCurrentIndex( this->index( rowCount() - 1, 1 ) );
+      treeView->selectionModel()->select( this->index( rowCount() - 1, 1 ), QItemSelectionModel::Rows );
+    }
+    else
+    {
+      moveBookmark( projectTableModel, qgisTableModel, index.row() - qgisTableModel.rowCount() );
+      treeView->scrollTo( this->index( qgisTableModel.rowCount() - 1, 1 ) );
+      treeView->setCurrentIndex( this->index( qgisTableModel.rowCount() - 1, 1 ) );
+      treeView->selectionModel()->select( this->index( qgisTableModel.rowCount() - 1, 1 ), QItemSelectionModel::Rows );
+    }
+    return true;
+  }
+  if ( index.column() < qgisTableModel.rowCount() )
+  {
+    return qgisTableModel.setData( index, value, role );
+  }
+  else
+  {
+    return projectTableModel.setData( this->index( index.row() - qgisTableModel.rowCount(), index.column() ), value, role );
+  }
+}
+
+Qt::ItemFlags QgsMergedBookmarksTableModel::flags( const QModelIndex& index ) const
+{
+  Q_UNUSED( index );
+
+  Qt::ItemFlags flags = Qt::ItemIsSelectable | Qt::ItemIsEnabled;
+  if ( index.column() == qgisTableModel.columnCount() )
+  {
+    if ( !projectOpen )
+    {
+      return Qt::ItemIsSelectable;
+    }
+    flags |= Qt::ItemIsUserCheckable;
+  }
+  else
+  {
+    flags |= Qt::ItemIsEditable;
+  }
+  return flags;
+}
+
+bool QgsMergedBookmarksTableModel::removeRows( int row, int count, const QModelIndex& parent )
+{
+  Q_ASSERT( count == 1 );
+
+  if ( row < qgisTableModel.rowCount() )
+  {
+    return qgisTableModel.removeRows( row, count, parent );
+  }
+  else
+  {
+    return projectTableModel.removeRows( row - qgisTableModel.rowCount(), count, parent );
+  }
+}
+
+QVariant QgsMergedBookmarksTableModel::headerData( int section, Qt::Orientation orientation, int role ) const
+{
+  if ( section == qgisTableModel.columnCount() )
+  {
+    if ( role == Qt::DisplayRole )
+    {
+      return QVariant( tr( "In Project" ) );
+    }
+    else
+    {
+      return qgisTableModel.headerData( 0, orientation, role );
+    }
+  }
+  else
+  {
+    return qgisTableModel.headerData( section, orientation, role );
+  }
+}
+
+QAbstractTableModel* QgsMergedBookmarksTableModel::qgisModel()
+{
+  return &qgisTableModel;
+}
+
+void QgsMergedBookmarksTableModel::moveBookmark( QAbstractTableModel& modelFrom, QAbstractTableModel& modelTo, int row )
+{
+  QSqlTableModel* mQgisModel = dynamic_cast<QSqlTableModel*>( &modelTo );
```

Comment:
should be qgisModel since it is a local variable here (probably a search / replace mistake)